File: scripts/fetch_latest_release.js
/**
 * GitHub Release MBTiles Fetcher
 * 
 * This script demonstrates how to fetch MBTiles files from GitHub Releases
 * in a web application.
 */

// Configuration
const REPO_OWNER = 'gabriel-briffe';
const REPO_NAME = 'arome';
const API_URL = `https://api.github.com/repos/${REPO_OWNER}/${REPO_NAME}/releases`;

/**
 * Fetch the latest release data from GitHub
 * @returns {Promise<Object>} The release data
 */
async function fetchLatestRelease() {
  try {
    const response = await fetch(`${API_URL}/latest`);
    if (!response.ok) {
      throw new Error(`Failed to fetch latest release: ${response.status} ${response.statusText}`);
    }
    return await response.json();
  } catch (error) {
    console.error('Error fetching latest release:', error);
    throw error;
  }
}

/**
 * Get a specific release by date
 * @param {string} date - Date in YYYY-MM-DD format
 * @returns {Promise<Object>} The release data
 */
async function fetchReleaseByDate(date) {
  try {
    // First get all releases
    const response = await fetch(API_URL);
    if (!response.ok) {
      throw new Error(`Failed to fetch releases: ${response.status} ${response.statusText}`);
    }
    
    const releases = await response.json();
    
    // Find the release with the matching tag
    const tagName = `arome-${date}`;
    const release = releases.find(r => r.tag_name === tagName);
    
    if (!release) {
      throw new Error(`No release found for date: ${date}`);
    }
    
    return release;
  } catch (error) {
    console.error(`Error fetching release for date ${date}:`, error);
    throw error;
  }
}

/**
 * Get all MBTiles files from a release
 * @param {Object} releaseData - The release data from GitHub API
 * @returns {Array<Object>} Array of file objects with name and url
 */
function getMBTilesFromRelease(releaseData) {
  return releaseData.assets
    .filter(asset => asset.name.endsWith('.mbtiles'))
    .map(asset => ({
      name: asset.name,
      url: asset.browser_download_url,
      size: asset.size,
      created_at: asset.created_at
    }));
}

/**
 * Get MBTiles file URLs for a specific pressure level
 * @param {Array<Object>} mbtiles - Array of MBTiles file objects
 * @param {number} pressure - Pressure level (e.g., 850)
 * @returns {Array<Object>} Filtered array of MBTiles for the specified pressure
 */
function getMBTilesForPressure(mbtiles, pressure) {
  const pressureStr = pressure.toString();
  return mbtiles.filter(file => file.name.includes(`_${pressureStr}.mbtiles`));
}

/**
 * Get MBTiles file URLs for a specific hour
 * @param {Array<Object>} mbtiles - Array of MBTiles file objects
 * @param {number} hour - Hour (0-23)
 * @returns {Array<Object>} Filtered array of MBTiles for the specified hour
 */
function getMBTilesForHour(mbtiles, hour) {
  const hourStr = hour.toString().padStart(2, '0');
  return mbtiles.filter(file => file.name.includes(`_${hourStr}_`));
}

/**
 * Example usage in a web application
 */
async function loadAROMEDataForDate(date = null) {
  try {
    // Get release data (latest or for specific date)
    const releaseData = date 
      ? await fetchReleaseByDate(date)
      : await fetchLatestRelease();
    
    console.log(`Loaded release: ${releaseData.name}`);
    
    // Get all MBTiles from the release
    const allMBTiles = getMBTilesFromRelease(releaseData);
    console.log(`Found ${allMBTiles.length} MBTiles files`);
    
    // Example: Filter for a specific pressure level and hour
    const pressure850MBTiles = getMBTilesForPressure(allMBTiles, 850);
    const hour12MBTiles = getMBTilesForHour(allMBTiles, 12);
    
    console.log(`Files for 850 hPa: ${pressure850MBTiles.length}`);
    console.log(`Files for 12:00: ${hour12MBTiles.length}`);
    
    // Return all the data
    return {
      releaseData,
      allMBTiles,
      pressure850MBTiles,
      hour12MBTiles
    };
  } catch (error) {
    console.error('Failed to load AROME data:', error);
    throw error;
  }
}

// Example of how to use this code in a web application:
/*
document.addEventListener('DOMContentLoaded', async () => {
  try {
    // Get today's date in YYYY-MM-DD format
    const today = new Date().toISOString().split('T')[0];
    
    // Try to load today's data or fall back to latest
    try {
      const data = await loadAROMEDataForDate(today);
      console.log(`Loaded data for ${today}`);
    } catch (error) {
      console.warn(`No data for ${today}, falling back to latest`);
      const data = await loadAROMEDataForDate();
    }
    
    // Use the data to populate your UI
    // ...
  } catch (error) {
    console.error('Error loading AROME data:', error);
    // Show error in UI
  }
});
*/ 

File: scripts/fetch_latest_release.py
#!/usr/bin/env python
"""
GitHub Release MBTiles Fetcher (Python Version)

This script demonstrates how to fetch MBTiles files from GitHub Releases
using Python requests.

Usage:
  python fetch_latest_release.py [--date YYYY-MM-DD] [--output-dir ./output]
"""

import os
import sys
import argparse
import requests
from datetime import datetime
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# Configuration
REPO_OWNER = 'gabriel-briffe'
REPO_NAME = 'arome'
API_URL = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/releases"

def fetch_latest_release():
    """Fetch the latest release data from GitHub"""
    try:
        response = requests.get(f"{API_URL}/latest")
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        logger.error(f"Error fetching latest release: {e}")
        raise

def fetch_release_by_date(date):
    """
    Get a specific release by date
    
    Parameters:
    -----------
    date : str
        Date in YYYY-MM-DD format
    
    Returns:
    --------
    dict
        The release data
    """
    try:
        # First get all releases
        response = requests.get(API_URL)
        response.raise_for_status()
        releases = response.json()
        
        # Find the release with the matching tag
        tag_name = f"arome-{date}"
        release = next((r for r in releases if r.get('tag_name') == tag_name), None)
        
        if not release:
            raise ValueError(f"No release found for date: {date}")
        
        return release
    except requests.exceptions.RequestException as e:
        logger.error(f"Error fetching release for date {date}: {e}")
        raise
    except ValueError as e:
        logger.error(str(e))
        raise

def get_mbtiles_from_release(release_data):
    """
    Get all MBTiles files from a release
    
    Parameters:
    -----------
    release_data : dict
        The release data from GitHub API
    
    Returns:
    --------
    list
        List of dictionaries with name, url, size, and created_at for each MBTiles file
    """
    return [
        {
            'name': asset['name'],
            'url': asset['browser_download_url'],
            'size': asset['size'],
            'created_at': asset['created_at']
        }
        for asset in release_data['assets']
        if asset['name'].endswith('.mbtiles')
    ]

def get_mbtiles_for_pressure(mbtiles, pressure):
    """
    Get MBTiles file info for a specific pressure level
    
    Parameters:
    -----------
    mbtiles : list
        List of MBTiles file dictionaries
    pressure : int
        Pressure level (e.g., 850)
    
    Returns:
    --------
    list
        Filtered list of MBTiles for the specified pressure
    """
    pressure_str = str(pressure)
    return [file for file in mbtiles if f"_{pressure_str}." in file['name']]

def get_mbtiles_for_hour(mbtiles, hour):
    """
    Get MBTiles file info for a specific hour
    
    Parameters:
    -----------
    mbtiles : list
        List of MBTiles file dictionaries
    hour : int
        Hour (0-23)
    
    Returns:
    --------
    list
        Filtered list of MBTiles for the specified hour
    """
    hour_str = f"{hour:02d}"
    return [file for file in mbtiles if f"_{hour_str}_" in file['name']]

def download_file(url, output_path):
    """
    Download a file from a URL to the specified path
    
    Parameters:
    -----------
    url : str
        URL of the file to download
    output_path : str
        Path to save the file to
    
    Returns:
    --------
    bool
        True if successful, False otherwise
    """
    try:
        logger.info(f"Downloading {url} to {output_path}")
        response = requests.get(url, stream=True)
        response.raise_for_status()
        
        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        
        with open(output_path, 'wb') as f:
            for chunk in response.iter_content(chunk_size=8192):
                f.write(chunk)
        
        logger.info(f"Successfully downloaded to {output_path}")
        return True
    except requests.exceptions.RequestException as e:
        logger.error(f"Error downloading file: {e}")
        return False
    except IOError as e:
        logger.error(f"IO Error: {e}")
        return False

def load_arome_data_for_date(date=None, output_dir=None):
    """
    Load AROME data for a specific date or the latest available
    
    Parameters:
    -----------
    date : str, optional
        Date in YYYY-MM-DD format, or None for latest
    output_dir : str, optional
        Directory to download files to, or None to just get info
    
    Returns:
    --------
    dict
        Dictionary with release data and MBTiles information
    """
    try:
        # Get release data (latest or for specific date)
        if date:
            release_data = fetch_release_by_date(date)
        else:
            release_data = fetch_latest_release()
            # Extract date from tag name
            date = release_data['tag_name'].replace('arome-', '')
        
        logger.info(f"Loaded release: {release_data['name']}")
        
        # Get all MBTiles from the release
        all_mbtiles = get_mbtiles_from_release(release_data)
        logger.info(f"Found {len(all_mbtiles)} MBTiles files")
        
        if output_dir:
            # Create output directory structure
            date_output_dir = os.path.join(output_dir, date)
            os.makedirs(date_output_dir, exist_ok=True)
            
            # Download all files
            for mbtile in all_mbtiles:
                output_path = os.path.join(date_output_dir, mbtile['name'])
                download_file(mbtile['url'], output_path)
        
        # Return all the data
        return {
            'release_data': release_data,
            'date': date,
            'all_mbtiles': all_mbtiles
        }
    except Exception as e:
        logger.error(f"Failed to load AROME data: {e}")
        raise

if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Fetch MBTiles from GitHub Releases"
    )
    parser.add_argument("--date", help="Date in YYYY-MM-DD format, or 'latest' for most recent")
    parser.add_argument("--output-dir", default="./downloaded_mbtiles", 
                        help="Directory to save downloaded files")
    
    args = parser.parse_args()
    
    try:
        # Use today's date if not specified
        if not args.date:
            args.date = "latest"
            
        if args.date.lower() == "latest":
            data = load_arome_data_for_date(output_dir=args.output_dir)
        else:
            data = load_arome_data_for_date(date=args.date, output_dir=args.output_dir)
        
        logger.info(f"Successfully processed release from {data['date']}")
        logger.info(f"Files downloaded to {args.output_dir}/{data['date']}/")
        
    except Exception as e:
        logger.error(f"Error: {e}")
        sys.exit(1)
    
    sys.exit(0) 